### DISCUSSION TITLE
Serverpod Not Signing in using Google on Android

### DISCUSSION THREAD

**Discussion by kaljitism** [serverpod/serverpod#3096] [Q&A]
I have followed @Isakdl 's Medium Blog, Official Docs, and looked at all Open/Closed Issues, but Google Sign in is just not working. Please Note that I am only using Android. 

**To Reproduce**
What I am doing - 
1. I have added `  serverpod_auth_server: 1.2.6` in server's pubspec, `pod.webServer.addRoute(auth.RouteGoogleSignIn(), '/googlesignin');` in server.dart and ran serverpod generate, applied migrations and ran the web server. 
2. I have defined `modules:
  serverpod_auth:
    nickname: auth` in generator.yaml
3. I have added `serverpod_auth_client: 1.2.6` in Client's pubspec
4. I have added `serverpod_auth_shared_flutter: 1.2.6` and `serverpod_auth_google_flutter: 1.2.6` in Flutter App's pubspec. 
5. I have created a project in GCP and enabled People API
6. I have created Client ID for Web Application - Defined Authorized JavaScript origins as `http://localhost:8082` and `http://localhost:49660`. Defined Authorized redirect URIs as `http://localhost:8082`, `http://localhost:49660`, `http://localhost:8082/googlesignin`.
7. I have downloaded the Web Application JSON and stored it as `server/config/google_client_secret.json`.
8. I have created Client ID for Android and defined SHA-1 certificate fingerprint. I have downloaded the Json file and stored it as `talkinbird_flutter/android/app/google-services.json`.
9. I am testing on emulator so my PC and my AVD both has the same IP. 
10. I have tried using my own IP Address and 10.0.0.2 - Using my own IP Address takes me to OAuth Consent Screen and then give me error, Using 10.0.0.2 straight away gives me error after clicking on my gmail id.
11.  I have tried using both Android's Client ID and Web Application's Client ID in SignInWithGoogleButton and found out that the Web Application one takes me to OAuth Consent Screen so I belive its the right one to use. 
12. I have tried running  './gradlew clean', `flutter clean` and `flutter pub cache repair` and Invalidate Cache and Restart in Android Studio multiple times in between trying various solutions as mentioned above. 

main.dart - 
```
late SessionManager sessionManager;
late Client client;

void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  const ipAddress = <My IP Address>;
  // const ipAddress = '10.0.0.2';

  client = Client(
    'http://$ipAddress:8080/',
    authenticationKeyManager: FlutterAuthenticationKeyManager(),
  )..connectivityMonitor = FlutterConnectivityMonitor();

  sessionManager = SessionManager(
    caller: client.modules.auth,
  );
  await sessionManager.initialize();

  runApp(const TalkinBird());
}
```

Sign In Screen - 
```
import 'package:flutter/material.dart';
import 'package:serverpod_auth_google_flutter/serverpod_auth_google_flutter.dart';
import 'package:talkinbird_flutter/screens/user_details_ui.dart';

import '../main.dart';

const _serverClientId = "<Web Application's Client ID>";

class SignInPage extends StatelessWidget {
  const SignInPage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Dialog(
          child: Container(
            width: 260,
            padding: const EdgeInsets.all(16),
            child: Column(
              mainAxisSize: MainAxisSize.min,
              crossAxisAlignment: CrossAxisAlignment.stretch,
              children: [
                SignInWithGoogleButton(
                  onSignedIn: () {
                    print('Signed in with Google!');
                    Navigator.of(context).push(MaterialPageRoute(
                      builder: (context) => const UserDetailsUI(),
                    ));
                  },
                  caller: client.modules.auth,
                  serverClientId: _clientId,
                  redirectUri: Uri.parse('http://localhost:8082/googlesignin'),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```

Error Logs right after the App is started on Android Emulator - 
```
Syncing files to device sdk gphone x86...
D/ProfileInstaller(20285): Installing profile for com.kaaljit.talkinbird
I/flutter (20285): Failed call: serverpod_auth.status.getUserInfo
I/flutter (20285): SocketException: HTTP connection timed out after 0:00:20.000000, host: <My IP Address>, port: 8080
I/flutter (20285): Failed call: user.getUser
I/flutter (20285): SocketException: HTTP connection timed out after 0:00:20.000000, host: <My IP Address>, port: 8080
[log] SocketException: HTTP connection timed out after 0:00:20.000000, host: <My IP Address>, port: 8080
I/flutter (20285): serverpod_auth_google: GoogleSignIn
I/flutter (20285): serverpod_auth_google: Signing out from google
I/flutter (20285): serverpod_auth_google: PlatformException(status, Failed to disconnect., null, null)
I/flutter (20285): Failed call: serverpod_auth.google.authenticateWithServerAuthCode
I/flutter (20285): SocketException: HTTP connection timed out after 0:00:20.000000, host: <My IP Address>, port: 8080
I/flutter (20285): serverpod_auth_google: SocketException: HTTP connection timed out after 0:00:20.000000, host: <My IP Address>, port: 8080
I/flutter (20285): #0      _ConnectionTarget.connect.<anonymous closure>.<anonymous closure> (dart:_http/http_impl.dart:2530:11)
I/flutter (20285): #1      _FutureListener.handleError (dart:async/future_impl.dart:180:22)
I/flutter (20285): #2      Future._propagateToListeners.handleError (dart:async/future_impl.dart:850:47)
I/flutter (20285): #3      Future._propagateToListeners (dart:async/future_impl.dart:871:13)
I/flutter (20285): #4      Future._completeError (dart:async/future_impl.dart:651:5)
I/flutter (20285): #5      Future.timeout.<anonymous closure> (dart:async/future_impl.dart:923:17)
I/flutter (20285): #6      Timer._createTimer.<anonymous closure> (dart:async-patch/timer_patch.dart:18:15)
I/flutter (20285): #7      _Timer._runTimers (dart:isolate-patch/timer_impl.dart:398:19)
I/flutter (20285): #8      _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:429:5)
I/flutter (20285): #9      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

**Expected behavior**
onSignedIn callback gets executed and Navigator takes me to the next screen as defined. 

**Serverpod versions**
 -  _CLI Version:_ 1.2.6
 - _Library version:_ 1.2.6

**Platform information**
I am on a Windows Machine, using Android Studio and working on Pixel 8 API 30 Android Emulator.
I have defined `org.gradle.jvmargs=-XX:MaxHeapSize=256m -Xmx2536m` in `gradle.properties` and multidex support is enabled. 
In my app/gradle - compileSdk is 34, ndkVersion "25.1.8937393", minSdkVersion is `flutter.minSdkVersion`, targetSdkVersion is `flutter.targetSdkVersion`

Here is the flutter doctor's outopt - 
```
PS C:\talkinbird\talkinbird_flutter> flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 3.19.4, on Microsoft Windows [Version 10.0.22621.3296], locale en-IN)
[√] Windows Version (Installed version of Windows is version 10 or higher)
[√] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[√] Chrome - develop for the web
[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.7.6)
[√] Android Studio (version 2023.2)
[√] Connected device (4 available)
[√] Network resources

• No issues found!
```

**Comment by centmeteenvin** [comment]
Possible issues I see,
In the signin button you pass _clientid serverClientId and you are giving the clientId parameter. Also it seems that there is an error in the connection between the server and the client. I am not familiar enough with the android emulator and how it handles networking. You might need to investigate this further. But if passing the 10.0.0.2 ip address is valid you might also need to add this to the allowed hosts and redirect URLs in your gcp project.

**Comment by kaljitism** [comment]
According to documentation, clientID is not used in Android and I am passing the right arguement, I have named it clientID because that's what it's named in GCP. 

It's clearly mentioned in the example directory and in some issues/ stackoverflow links that you have to use the IP address of your system. 

10.0.0.2 doesn't work. Also, if I go ahead and try to add http://10.0.0.2:8082/ in redirect URL, it doesn't work. Adding http://<Any IP Add>:8082/ also doesn't work. GCP Console says "Invalid Origin: must end with a public top-level domain (such as .com or .org)" which doesn't seem to be the case with localhost urls. 

I also tried testing on real physical device on the same IP as my system and that doesn't work either.

**Comment by centmeteenvin** [comment]
I have read up a bit on the android emulator [here](https://stackoverflow.com/questions/9808560/why-do-we-use-10-0-2-2-to-connect-to-local-web-server-instead-of-using-computer.) and it tells me the ip address of the emulator is 10.0.2.2 instead of 10.0.0.2. I actually think you got further in the process when you used the computer's local ip address.  I hope this resolves that issue. I don't know how it will behave with regards to the GCP allowed origins for which I will need more information.

If my information of the 10.0.2.0/24 address space is outdated can you show me when it has changed.
Then I think if the 10.0.2.2 ip address doesn't work try again with the local ip and send logs of both the flutter and serverpod application

**Comment by kaljitism** [comment]
Read my Point No. 10 at the top level comment to see the difference between using real IP address and 10.0.02 . Even if it works at 10.0.0.2, it doesnt resolve my issue since the app is intend to work on real device. 

These are the places its mentioned to use a real IP for android. These images are from [Part 1](https://medium.com/serverpod/getting-started-with-serverpod-authentication-part-1-72c25280e6e9) and [Part 2](https://medium.com/serverpod/integrating-google-sign-in-with-serverpod-authentication-part-2-6fade3099baf) of the article mentioned in docs. 

![1](https://github.com/serverpod/serverpod/assets/52685389/cca8f0c1-2672-4db6-b3e6-ff064abf1d0c)

![2](https://github.com/serverpod/serverpod/assets/52685389/654656e4-4290-4346-a566-dcf31b0400e2)

**Comment by centmeteenvin** [comment]
You haven't read my reply thoroughly. I indicate that 10.0.0.2 is not in the IP address space of the android emulator. 10.0.2.0/24 is with the default gateway at 10.0.2.2 which is connected to your desktop. Il put them below eachother:
10.0.0.2 (the one you are currently using)
10.0.2.2 (the one you should be using according to the documentation I found)

Either way I now suggest you use the actual ip-address of the computer (the Ip-address on your local network) instead of the 10.0.2.2 as it more closely resembles the real world use case. This should resolve the http time out errors. They were probably caused by the IP you gave not being connected to an actual device in the virtual network.

To resolve the other issues we need more data. I will need the logs of your flutter application and your serverpod backend when you make the calls.

**Comment by SandPod** [comment]
Converting this to a discussion. Was the problem resolved @kaljitism?

**Comment by kaljitism** [comment]
Yeah, it was solved over time.. Thaks for replying!
